### ISSUE TITLE
QT app voice uses grandma version

### ISSUE THREAD

**Issue by ArkadiuszNiemiec** [brailcom/speechd#627]
I am using a `QGroundControl`, QT based app, and have just moved to Fedora 34. On the new system, the voice sounded odd, like a grandpa speaking to me. I digged a little and made sure that the QT app sets locale correctly, and yes:
https://github.com/mavlink/qgroundcontrol/blob/b18d2fdacef23815c6961e982f3d9fac06879357/src/Audio/AudioOutput.cc#L33-35

Then I digged a little more and this is what I have found in the log of `speechd`:

```
tail -f /run/user/1000/speech-dispatcher/log/speech-dispatcher.log | grep grand
[Tue Oct 26 21:22:59 2021 : 125478] speechd:     21:DATA:|SET SELF SYNTHESIS_VOICE English (America)+grandma
synthesis_voice=English (America)+grandma
```

`spd-say` works correctly.

What's wrong? Is that some kind of Easter egg? :smile: 

I have rebuilt the `QGroundControl` with locale forced to `pl_PL` and this was in the logs:
```
[Tue Oct 26 22:06:28 2021 : 330969] speechd:     21:DATA:|SET SELF SYNTHESIS_VOICE Polish+grandma
```

**Comment by ArkadiuszNiemiec** [comment]
@sthibaul I have just installed fresh Ubuntu 21.10 that has speech-dispatcher 0.10.2 and the same happened. It's not a Fedora issue, the grandma will show in every app on new distributions if the developer will set only language. I have found at least 3 QT apps that has this problem.

**Comment by ArkadiuszNiemiec** [comment]
I just tried to do a PR for the `QGroundControl`. To set the voice I have to ask QT for all voices and then select a specific one by name. This is a nightmare to do just for a specific voice engine only on Linux. 
But I confirmed, grandma is first on the list. 

I am closing the issue as this is a `espeak-ng` issue. Thank you.

**Comment by sthibaul** [comment]
I don't think this is an espeak-ng problem, but rather a misunderstanding between libspeechd and qtspeech.

> grandma is first on the list.

That's the actual problem. Why is it so? speech-dispatcher is not providing it first in the voice list, on the contrary actually:

```
249-English (America)	en-US	none
249-English (America)+Andrea	en-US	Andrea
249-English (America)+Annie	en-US	Annie
249-English (America)+Antonio	en-US	Antonio
249-English (America)+Auntie	en-US	Auntie
[...]
249-English (America)+male8	en-US	male8
249-English (America)+female1	en-US	female1
249-English (America)+croak	en-US	croak
249-English (America)+male1	en-US	male1
249-English (America)+grandma	en-US	grandma
```

Possibly qtspeech is simply reversing the list, and that's the mere fix to be done?

**Comment by ArkadiuszNiemiec** [comment]
I run:
```
m_voices = m_speech->availableVoices();
for (const QVoice &voice : qAsConst(m_voices)) {
    auto test = QString("%1 - %2 - %3").arg(voice.name())
            .arg(QVoice::genderName(voice.gender()))
            .arg(QVoice::ageName(voice.age()));
    qInfo() << test;
}
```

And the output is:
```
"English (America)+grandma - Unknown Gender - Other Age"
"English (America)+male1 - Unknown Gender - Other Age"
"English (America)+croak - Unknown Gender - Other Age"
"English (America)+female1 - Unknown Gender - Other Age"
"English (America)+male8 - Unknown Gender - Other Age"
...
"English (America)+Antonio - Unknown Gender - Other Age"
"English (America)+Annie - Unknown Gender - Other Age"
"English (America)+Andrea - Unknown Gender - Other Age"
"English (America) - Unknown Gender - Other Age"
"English (Caribbean)+grandma - Unknown Gender - Other Age"
"English (Caribbean)+male1 - Unknown Gender - Other Age"
"English (Caribbean)+croak - Unknown Gender - Other Age"
...
"English (Caribbean)+Auntie - Unknown Gender - Other Age"
"English (Caribbean)+Antonio - Unknown Gender - Other Age"
"English (Caribbean)+Annie - Unknown Gender - Other Age"
"English (Caribbean)+Andrea - Unknown Gender - Other Age"
"English (Caribbean) - Unknown Gender - Other Age"
```

**Comment by jpwhiting** [comment]
Yeah, that's definitely possible. Here's how QtSpeech gets the list of
voices and modules, etc. currently (I'll have to check to see what QtSpeech
5.12.7 does, this is latest, so 5.15.2 or whatever.

void QTextToSpeechEngineSpeechd::updateVoices()
{
    // First get the list of modules
    char **modules = spd_list_modules(speechDispatcher);
#ifdef HAVE_SPD_090
    // If spd is new enough get the current module from it
    char *original_module = spd_get_output_module(speechDispatcher);
#else
    // Otherwise just use the first one given back
    char *original_module = modules[0];
#endif
    QVoice originalVoice;
    char **module = modules;
    while (module != nullptr && module[0] != nullptr) {
        // Tell spd to use the module
        spd_set_output_module(speechDispatcher, module[0]);

        // Then get the list of voices from it
        SPDVoice **voices = spd_list_synthesis_voices(speechDispatcher);
        int i = 0;
        // Iterate the list of voices
        while (voices != nullptr && voices[i] != nullptr) {
            QLocale locale = localeForVoice(voices[i]);
            // Add that voice's locale to our known locales if not there
already
            if (!m_locales.contains(locale))
                m_locales.append(locale);
            const QString name = QString::fromUtf8(voices[i]->name);
            // iterate over genders and ages, creating a voice for each one
            QVoice voice = createVoice(name, QVoice::Unknown,
QVoice::Other, QLatin1String(module[0]));
            m_voices.insert(locale.name(), voice);
            // If the current module is the "original_module" from above,
remember the current voice as default
            if (module[0] == original_module && i == 0) {
                // in lack of better options, remember the first voice as
default
                originalVoice = voice;
            }
            ++i;
        }
        // free voices.
#ifdef HAVE_SPD_090
        free_spd_voices(voices);
#endif
        ++module;
    }

#ifdef HAVE_SPD_090
    // Also free modules.
    free_spd_modules(modules);
#endif

Since it's just setting the voice to the "first" from the original_module
that likely is getting the wrong default. It probably should instead if spd
supports it get the originalVoice from spd itself. But I don't see an spd
api to get the current or default voice. I'm looking at master branch of
speechd repo. Probably need to add one. Or if that proves tricky at least
give the list from spd_list_synthesis_voices in the same order as spd-say
gives it in. All that said a change in QtSpeech isn't likely to come soon,
it hasn't yet been included in Qt 6 releases, should be in 6.3 as far as I
understand it, so a fix could go in there. Alternatively we could promote a
patch to Qt 5.15.2 in linux distributions if needed and hope packagers in
those distributions add it to the patches they add already if any.

BR,
Jeremy

**Comment by sthibaul** [comment]
Taking the first language returned by speech-dispatcher should just all fine. Please just fix the qtspeech code to avoid reverting the list order.

**Comment by jpwhiting** [comment]
I don't see anywhere in there that's reversing the order. What am I missing?

**Comment by sthibaul** [comment]
> m_voices.insert(locale.name(), voice);

That's very probably adding at the beginning of the array, thus reverting the list.

**Comment by jpwhiting** [comment]
I see, it's a multimap by locale, that's messing up the order for sure:

    // Voices mapped by their locale name.
    QMultiMap<QString, QVoice> m_voices;

Will get a fix ready for distros to add and a fix ready for Qt 6.3. Thanks
for your patience and explanation Samuel.

BR,
Jeremy

**Comment by sthibaul** [comment]
Ok thanks!

**Comment by jpwhiting** [comment]
Found it. After the above code we set the voice to the originalVoice, but
later when the locale is set (which it is in almost all Qt applications
using QtSpeech from what I saw) we use the "first" voice from the multimap
but we use MultiMap::values(locale) to get the list, which returns the list
in order of most recently added to oldest added. Which gives the list in
reverse. Reversing the reverse order list we get before selecting the first
one should fix that. I'll give it a try this evening and get patches for Qt
5.12.7 and 5.15.2 ready. Will send those to the Qt a11y list and mention it
when distros reach out if any. Feel free to point it out to the distros you
use. I'll reply here with a link to the mailing list once that is ready and
sent, etc.

BR,
Jeremy
